Objects/regions.c

```diff
@@ -9,13 +9,67 @@
 #include "pycore_interp.h"
 #include "pycore_object.h"
 #include "pycore_regions.h"
+#include "pycore_pyerrors.h"
+
+#define Py_REGION_VISITED_FLAG ((Py_uintptr_t)0x2)
+static inline Py_uintptr_t Py_REGION_WITH_FLAGS(PyObject *ob) {
+    return ob->ob_region;
+}
+#define Py_REGION_WITH_FLAGS(ob) Py_REGION_WITH_FLAGS(_PyObject_CAST(ob))
+#define REGION_SET_FLAG(ob, flag) (Py_SET_REGION(ob, Py_REGION_WITH_FLAGS(ob) | flag))
+#define REGION_GET_FLAG(ob, flag) (Py_REGION_WITH_FLAGS(ob) & flag)
+#define REGION_CLEAR_FLAG(ob, flag) (Py_SET_REGION(ob, Py_REGION_WITH_FLAGS(ob) & (~flag)))
 
 typedef struct regionmetadata regionmetadata;
 typedef struct PyRegionObject PyRegionObject;
 
 static PyObject *PyRegion_add_object(PyRegionObject *self, PyObject *args);
 static PyObject *PyRegion_remove_object(PyRegionObject *self, PyObject *args);
 
+/**
+ * Global status for performing the region check.
+ */
+bool do_region_check = false;
+
+// The src object for an edge that invalidated the invariant.
+PyObject* error_src = Py_None;
+
+// The tgt object for an edge that invalidated the invariant.
+PyObject* error_tgt = Py_None;
+
+// Once an error has occurred this is used to surpress further checking
+bool error_occurred = false;
```

Comment:
This is my issue, but as you have touched it.  I wonder if this would be better:
```suggestion
// Once an error has occurred this is used to surpress further checking
bool invariant_failure_occurred = false;
```
Then it is clearly different from the runtimes ErrorOccurred.